File: app/src/main/java/com/nssf/datacapture/MrzRepairEngine.java
package com.nssf.datacapture;

import androidx.appcompat.app.AlertDialog;
import android.content.Context;
import android.graphics.Color;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

/**
 * MrzRepairEngine
 * 
 * Bounded single-character check-digit repair with user confirmation.
 * When 1 check digit fails, this engine generates candidate repairs by
 * substituting visually confusable characters (0↔O, 1↔I, etc.) at each
 * position in the failed field, then re-validates. Only candidates that
 * achieve HIGH confidence are presented to the user for manual confirmation.
 * 
 * NEVER auto-applies repairs. Silent data corruption is worse than a retry.
 */
public class MrzRepairEngine {

    private final Context context;
    private final RepairCallback callback;

    public interface RepairCallback {
        void onRepairConfirmed(CardRecord repairedRecord);
        void onRepairCancelled();
    }

    public MrzRepairEngine(Context context, RepairCallback callback) {
        this.context = context;
        this.callback = callback;
    }

    /**
     * Attempt to repair a MEDIUM-confidence record.
     * 
     * @param record The parsed record with ValidationConfidence.MEDIUM
     * @param line1 The raw MRZ line 1 (30 chars)
     * @param line2 The raw MRZ line 2 (30 chars)
     */
    public void attemptRepair(CardRecord record, String line1, String line2) {
        if (record == null || record.validationConfidence != UgandaIdParser.ValidationConfidence.MEDIUM) {
            if (callback != null) callback.onRepairCancelled();
            return;
        }

        // Determine which check digit failed
        List<RepairCandidate> candidates = generateCandidates(line1, line2);

        if (candidates.isEmpty()) {
            Toast.makeText(context, "⚠️ Could not auto-suggest a repair. Please verify manually.", 
                Toast.LENGTH_LONG).show();
            if (callback != null) callback.onRepairCancelled();
            return;
        }

        showRepairDialog(record, candidates, line1, line2);
    }

    /**
     * Generate repair candidates by trying single-character substitutions
     * on the field that failed check digit validation.
     */
    private List<RepairCandidate> generateCandidates(String line1, String line2) {
        List<RepairCandidate> candidates = new ArrayList<>();

        if (line1 == null || line1.length() < 14 || line2 == null || line2.length() < 14) {
            return candidates;
        }

        // Test each field individually
        String[][] fields = {
            {"Card Number", line1.substring(5, Math.min(14, line1.length())), "cd1", line1, line2, "1"},
            {"Date of Birth", line2.substring(0, Math.min(6, line2.length())), "cd2", line1, line2, "2"},
            {"Expiry Date", line2.length() >= 14 ? line2.substring(8, 14) : "", "cd3", line1, line2, "2"}
        };

        for (String[] field : fields) {
            String fieldName = field[0];
            String fieldValue = field[1];
            if (fieldValue.isEmpty()) continue;
            String cdType = field[2];
            String originalLine1 = field[3];
            String originalLine2 = field[4];
            String targetLine = field[5]; // "1" = line1, "2" = line2

            char[] chars = fieldValue.toCharArray();
            for (int i = 0; i < chars.length; i++) {
                char original = chars[i];
                List<Character> alternatives = getOcrAlternatives(original);

                for (char alt : alternatives) {
                    chars[i] = alt;
                    String repairedValue = new String(chars);

                    // Reconstruct the full line with the repaired value
                    String testLine1 = originalLine1;
                    String testLine2 = originalLine2;

                    if ("1".equals(targetLine)) {
                        testLine1 = originalLine1.substring(0, 5) + repairedValue + originalLine1.substring(14);
                    } else if ("cd2".equals(cdType)) {
                        testLine2 = repairedValue + originalLine2.substring(6);
                    } else if ("cd3".equals(cdType)) {
                        testLine2 = originalLine2.substring(0, 8) + repairedValue + originalLine2.substring(14);
                    }

                    // Re-validate
                    UgandaIdParser.ValidationResult result = 
                        UgandaIdParser.validateCheckDigits(testLine1, testLine2);

                    if (result.confidence == UgandaIdParser.ValidationConfidence.HIGH) {
                        candidates.add(new RepairCandidate(
                            fieldName,
                            i,
                            original,
                            alt,
                            repairedValue,
                            testLine1,
                            testLine2,
                            "Changed '" + original + "' to '" + alt + "' at position " + (i + 1) + " in " + fieldName
                        ));
                    }
                }
                chars[i] = original; // restore
            }
        }

        return candidates;
    }

    /**
     * Get visually confusable OCR alternatives for a character.
     */
    private List<Character> getOcrAlternatives(char c) {
        List<Character> alts = new ArrayList<>();

        switch (c) {
            case '0': alts.add('O'); break;
            case 'O': alts.add('0'); break;
            case '1': alts.add('I'); alts.add('L'); break;
            case 'I': alts.add('1'); alts.add('L'); break;
            case 'L': alts.add('1'); alts.add('I'); break;
            case '5': alts.add('S'); break;
            case 'S': alts.add('5'); break;
            case '8': alts.add('B'); break;
            case 'B': alts.add('8'); break;
            case '6': alts.add('G'); break;
            case 'G': alts.add('6'); break;
            case '4': alts.add('A'); break;
            case 'A': alts.add('4'); break;
            case '2': alts.add('Z'); break;
            case 'Z': alts.add('2'); break;
            case '7': alts.add('T'); break;
            case 'T': alts.add('7'); break;
            case '9': alts.add('P'); break;
            case 'P': alts.add('9'); break;
            case '3': alts.add('E'); break;
            case 'E': alts.add('3'); break;
        }

        return alts;
    }

    /**
     * Show the repair confirmation dialog to the user.
     */
    private void showRepairDialog(CardRecord originalRecord, 
                                   List<RepairCandidate> candidates,
                                   String line1, String line2) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle("⚠️ MRZ Check Digit Warning");
        builder.setMessage("One field may have an OCR error. Please confirm the correct value:");

        LinearLayout layout = new LinearLayout(context);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(32, 16, 32, 16);

        // Show original failed value
        TextView originalLabel = new TextView(context);
        originalLabel.setText("Original scan: " + getFailedFieldSummary(originalRecord, line1, line2));
        originalLabel.setTextColor(Color.GRAY);
        originalLabel.setPadding(0, 0, 0, 16);
        layout.addView(originalLabel);

        // Show top 3 candidates as buttons
        int maxCandidates = Math.min(3, candidates.size());
        for (int i = 0; i < maxCandidates; i++) {
            RepairCandidate cand = candidates.get(i);
            Button btn = new Button(context);
            btn.setText((i + 1) + ". " + cand.description);
            btn.setTag(cand);
            btn.setOnClickListener(v -> {
                RepairCandidate selected = (RepairCandidate) v.getTag();
                CardRecord repaired = UgandaIdParser.parseMrzLines(
                    java.util.Arrays.asList(selected.repairedLine1, selected.repairedLine2, line2)
                );
                if (repaired != null) {
                    repaired.validationConfidence = UgandaIdParser.ValidationConfidence.HIGH;
                    repaired.validationFailures = 0;
                    // Preserve phone from original if already entered
                    repaired.phoneNumber = originalRecord.phoneNumber;
                    Toast.makeText(context, "✅ Repair applied! All check digits pass.", 
                        Toast.LENGTH_SHORT).show();
                    if (callback != null) callback.onRepairConfirmed(repaired);
                }
            });
            layout.addView(btn);
        }

        builder.setView(layout);
        builder.setNegativeButton("Enter Manually", (dialog, which) -> {
            if (callback != null) callback.onRepairCancelled();
        });
        builder.setCancelable(false);
        builder.show();
    }

    private String getFailedFieldSummary(CardRecord record, String line1, String line2) {
        if (line1 == null || line1.length() < 15 || line2 == null || line2.length() < 15) {
            return "MRZ check digit mismatch";
        }
        StringBuilder sb = new StringBuilder();

        String cardNum = line1.substring(5, 14);
        int expectedCd1 = UgandaIdParser.calculateCheckDigit(cardNum);
        int actualCd1 = Character.getNumericValue(line1.charAt(14));
        if (expectedCd1 != actualCd1) sb.append("Card# ").append(cardNum).append(" (CD fail) ");

        String dob = line2.substring(0, 6);
        int expectedCd2 = UgandaIdParser.calculateCheckDigit(dob);
        int actualCd2 = Character.getNumericValue(line2.charAt(6));
        if (expectedCd2 != actualCd2) sb.append("DOB ").append(dob).append(" (CD fail) ");

        String expiry = line2.substring(8, 14);
        int expectedCd3 = UgandaIdParser.calculateCheckDigit(expiry);
        int actualCd3 = Character.getNumericValue(line2.charAt(14));
        if (expectedCd3 != actualCd3) sb.append("Expiry ").append(expiry).append(" (CD fail)");

        return sb.toString().trim();
    }

    /**
     * Internal data class for repair candidates.
     */
    public static class RepairCandidate {
        public final String fieldName;
        public final int position;
        public final char originalChar;
        public final char replacementChar;
        public final String repairedValue;
        public final String repairedLine1;
        public final String repairedLine2;
        public final String description;

        public RepairCandidate(String fieldName, int position, char originalChar, 
                               char replacementChar, String repairedValue,
                               String repairedLine1, String repairedLine2, 
                               String description) {
            this.fieldName = fieldName;
            this.position = position;
            this.originalChar = originalChar;
            this.replacementChar = replacementChar;
            this.repairedValue = repairedValue;
            this.repairedLine1 = repairedLine1;
            this.repairedLine2 = repairedLine2;
            this.description = description;
        }
    }
}

File: js/mrz-repair-engine.js
/**
 * mrz-repair-engine.js
 * PWA Browser Bounded Check-Digit Repair Engine
 * 
 * When UgandaIdParser returns ValidationConfidence.MEDIUM (1 check digit failure),
 * this engine generates candidate repairs by substituting visually confusable
 * characters, then presents them to the user for confirmation via a modal dialog.
 * 
 * NEVER auto-applies repairs. User must explicitly confirm.
 * 
 * @version 1.0.0
 */

import { UgandaIdParser, ValidationConfidence } from './uganda-id-parser.js';

/**
 * OCR confusion table for single-character substitutions.
 */
const OCR_CONFUSIONS = {
    '0': ['O'], 'O': ['0'],
    '1': ['I', 'L'], 'I': ['1', 'L'], 'L': ['1', 'I'],
    '5': ['S'], 'S': ['5'],
    '8': ['B'], 'B': ['8'],
    '6': ['G'], 'G': ['6'],
    '4': ['A'], 'A': ['4'],
    '2': ['Z'], 'Z': ['2'],
    '7': ['T'], 'T': ['7'],
    '9': ['P'], 'P': ['9'],
    '3': ['E'], 'E': ['3']
};

/**
 * Generate repair candidates for a MEDIUM-confidence MRZ parse.
 * 
 * @param {string} line1 — 30-char MRZ line 1
 * @param {string} line2 — 30-char MRZ line 2
 * @returns {Array<Object>} Candidates that achieve HIGH confidence
 */
export function generateRepairCandidates(line1, line2) {
    const candidates = [];

    const fields = [
        { name: 'Card Number', value: line1.slice(5, 14), cd: 'cd1', line: '1', offset: 5 },
        { name: 'Date of Birth', value: line2.slice(0, 6), cd: 'cd2', line: '2', offset: 0 },
        { name: 'Expiry Date', value: line2.slice(8, 14), cd: 'cd3', line: '2', offset: 8 }
    ];

    for (const field of fields) {
        const chars = field.value.split('');
        for (let i = 0; i < chars.length; i++) {
            const original = chars[i];
            const alternatives = OCR_CONFUSIONS[original] || [];

            for (const alt of alternatives) {
                chars[i] = alt;
                const repairedValue = chars.join('');

                // Reconstruct lines
                let testLine1 = line1;
                let testLine2 = line2;

                if (field.line === '1') {
                    testLine1 = line1.slice(0, 5) + repairedValue + line1.slice(14);
                } else if (field.cd === 'cd2') {
                    testLine2 = repairedValue + line2.slice(6);
                } else if (field.cd === 'cd3') {
                    testLine2 = line2.slice(0, 8) + repairedValue + line2.slice(14);
                }

                // Re-validate
                const result = UgandaIdParser.validateCheckDigits(testLine1, testLine2);
                if (result.confidence === ValidationConfidence.HIGH) {
                    candidates.push({
                        fieldName: field.name,
                        position: i + 1,
                        originalChar: original,
                        replacementChar: alt,
                        repairedValue,
                        repairedLine1: testLine1,
                        repairedLine2: testLine2,
                        description: `Changed '${original}' to '${alt}' at position ${i + 1} in ${field.name}`
                    });
                }
            }
            chars[i] = original; // restore
        }
    }

    return candidates;
}

/**
 * Show repair confirmation modal in the browser.
 * 
 * @param {Object} originalRecord
 * @param {Array<Object>} candidates
 * @param {Function} onConfirm — called with repaired CardRecord
 * @param {Function} onCancel — called when user rejects all repairs
 */
export function showRepairDialog(originalRecord, candidates, onConfirm, onCancel) {
    const existing = document.getElementById('mrz-repair-modal');
    if (existing) existing.remove();

    const modal = document.createElement('div');
    modal.id = 'mrz-repair-modal';
    modal.style.cssText = `
        position: fixed; top: 0; left: 0; width: 100%; height: 100%;
        background: rgba(0,0,0,0.6); z-index: 10000;
        display: flex; align-items: center; justify-content: center;
        font-family: sans-serif;
    `;

    const panel = document.createElement('div');
    panel.style.cssText = `
        background: white; border-radius: 12px; padding: 24px;
        max-width: 90%; width: 400px; max-height: 80vh; overflow-y: auto;
        box-shadow: 0 8px 32px rgba(0,0,0,0.3);
    `;

    const title = document.createElement('h3');
    title.textContent = '⚠️ MRZ Check Digit Warning';
    title.style.marginTop = '0';
    panel.appendChild(title);

    const subtitle = document.createElement('p');
    subtitle.textContent = 'One field may have an OCR error. Please confirm the correct value:';
    subtitle.style.color = '#666';
    panel.appendChild(subtitle);

    const originalBox = document.createElement('div');
    originalBox.style.cssText = 'background: #f5f5f5; padding: 12px; border-radius: 8px; margin: 12px 0; font-size: 13px;';
    originalBox.innerHTML = `<strong>Original:</strong><br>NIN: ${originalRecord.nin}<br>Card: ${originalRecord.cardNumber}<br>DOB: ${originalRecord.dateOfBirth}`;
    panel.appendChild(originalBox);

    const maxShow = Math.min(3, candidates.length);
    for (let i = 0; i < maxShow; i++) {
        const cand = candidates[i];
        const btn = document.createElement('button');
        btn.textContent = `${i + 1}. ${cand.description}`;
        btn.style.cssText = `
            display: block; width: 100%; padding: 14px; margin: 8px 0;
            border: 2px solid #4CAF50; border-radius: 8px; background: #E8F5E9;
            color: #2E7D32; font-size: 14px; font-weight: 600; cursor: pointer;
        `;
        btn.onmouseenter = () => { btn.style.background = '#C8E6C9'; };
        btn.onmouseleave = () => { btn.style.background = '#E8F5E9'; };
        btn.onclick = () => {
            const repaired = UgandaIdParser.parseMrzLines([
                cand.repairedLine1,
                cand.repairedLine2,
                '<<'
            ]);
            if (repaired) {
                repaired.validationConfidence = ValidationConfidence.HIGH;
                repaired.validationFailures = 0;
                repaired.phoneNumber = originalRecord.phoneNumber;
            }
            modal.remove();
            if (onConfirm) onConfirm(repaired || originalRecord);
        };
        panel.appendChild(btn);
    }

    const manualBtn = document.createElement('button');
    manualBtn.textContent = 'Enter Manually';
    manualBtn.style.cssText = `
        display: block; width: 100%; padding: 14px; margin-top: 12px;
        border: 2px solid #F44336; border-radius: 8px; background: #FFEBEE;
        color: #C62828; font-size: 14px; font-weight: 600; cursor: pointer;
    `;
    manualBtn.onclick = () => {
        modal.remove();
        if (onCancel) onCancel();
    };
    panel.appendChild(manualBtn);

    modal.appendChild(panel);
    document.body.appendChild(modal);
}

/**
 * Convenience: attempt repair and auto-show dialog.
 * 
 * @param {Object} record — MEDIUM confidence record
 * @param {string} line1 — raw line 1
 * @param {string} line2 — raw line 2
 * @param {Function} onConfirm
 * @param {Function} onCancel
 */
export function attemptRepair(record, line1, line2, onConfirm, onCancel) {
    if (!record || record.validationConfidence !== ValidationConfidence.MEDIUM) {
        if (onCancel) onCancel();
        return;
    }

    const candidates = generateRepairCandidates(line1, line2);
    if (candidates.length === 0) {
        alert('⚠️ Could not auto-suggest a repair. Please verify manually.');
        if (onCancel) onCancel();
        return;
    }

    showRepairDialog(record, candidates, onConfirm, onCancel);
}
